diff --git a/index.js b/index.js
index c00a171bc668e0c1ee8bb872e530242515cd885c..821b21711c6bde2a3404a447566bac383cbe2582 100644
--- a/index.js
+++ b/index.js
@@ -11,36 +11,26 @@ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLI
 
 //  Dependencies
 // ================================================
-var pid = typeof process !== 'undefined' && process.pid ? process.pid.toString(36) : '' ;
-var address = '';
-if(typeof __webpack_require__ !== 'function' && typeof require !== 'undefined'){
-    var mac = '', os = require('os'); 
-    if(os.networkInterfaces) var networkInterfaces = os.networkInterfaces();
-    if(networkInterfaces){
-        loop:
-        for(let interface_key in networkInterfaces){
-            const networkInterface = networkInterfaces[interface_key];
-            const length = networkInterface.length;
-            for(var i = 0; i < length; i++){
-                if(networkInterface[i] !== undefined && networkInterface[i].mac && networkInterface[i].mac != '00:00:00:00:00:00'){
-                    mac = networkInterface[i].mac; break loop;
-                }
-            }
-        }
-        address = mac ? parseInt(mac.replace(/\:|\D+/gi, '')).toString(36) : '' ;
-    }
-} 
+const pid = typeof process !== 'undefined' && process.pid ? process.pid.toString(36) : '' ;
 
 //  Exports
 // ================================================
-module.exports = module.exports.default = function(prefix, suffix){ return (prefix ? prefix : '') + address + pid + now().toString(36) + (suffix ? suffix : ''); }
-module.exports.process = function(prefix, suffix){ return (prefix ? prefix : '') + pid + now().toString(36) + (suffix ? suffix : ''); }
-module.exports.time    = function(prefix, suffix){ return (prefix ? prefix : '') + now().toString(36) + (suffix ? suffix : ''); }
+export default function(prefix, suffix) {
+    return (prefix ? prefix : '') + pid + now().toString(36) + (suffix ? suffix : ''); 
+}
+
+export function process(prefix, suffix) {
+    return (prefix ? prefix : '') + pid + now().toString(36) + (suffix ? suffix : ''); 
+}
+
+export function time(prefix, suffix) {
+    return (prefix ? prefix : '') + now().toString(36) + (suffix ? suffix : '');
+}
 
 //  Helpers
 // ================================================
 function now(){
-    var time = Date.now();
-    var last = now.last || time;
+    const time = Date.now();
+    const last = now.last || time;
     return now.last = time > last ? time : last + 1;
 }
diff --git a/package.json b/package.json
index 8cbc36fccab255f7cad35f1d54ea276c5da3c80d..986d30dd527269c5f7d5fb8970e3da2e73a2d0ea 100644
--- a/package.json
+++ b/package.json
@@ -2,6 +2,7 @@
   "name": "uniqid",
   "version": "5.4.0",
   "description": "Unique ID Generator",
+  "type": "module",
   "homepage": "http://github.com/adamhalasz/uniqid/",
   "keywords": [
     "unique id",
@@ -26,7 +27,7 @@
     "email": "adam@aimform.com",
     "url": "http://adamhalasz.com/"
   },
-  "main": "index.js",
+  "exports": "./index.js",
   "dependencies": {},
   "devDependencies": {}
 }
